shell/platform/linux/public/flutter_linux/fl_engine.h

```diff
@@ -0,0 +1,57 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_LINUX_FL_ENGINE_H_
+#define FLUTTER_SHELL_PLATFORM_LINUX_FL_ENGINE_H_
+
+#if !defined(__FLUTTER_LINUX_INSIDE__) && !defined(FLUTTER_LINUX_COMPILATION)
+#error "Only <flutter_linux/flutter_linux.h> can be included directly."
+#endif
+
+#include <glib-object.h>
+
+#include "fl_dart_project.h"
+
+G_BEGIN_DECLS
+
+G_DECLARE_DERIVABLE_TYPE(FlEngine, fl_engine, FL, ENGINE, GObject)
+
+struct _FlEngineClass {
+  /*< private >*/
+  GObjectClass parent_class;
+
+  gboolean (*start)(FlEngine* engine, GError** error);
+  void* (*get_proc_address)(FlEngine* engine, const char* name);
```

Comment:
Ideally the engine object (which will be a public API surface, for running in headless mode) shouldn't have graphics details in the API. The way this works in the ObjC API is that the engine has an optional view property, and then internally it implements all these callbacks as forwarding them to the view if it's set, with fallbacks to essentially a no-op implementation (see https://github.com/flutter/engine/blob/master/shell/platform/darwin/macos/framework/Source/FlutterEngine.mm#L106 and the methods they call to).

I'd prefer to do that here as well, both because it makes the API surface of engine graphics-agnostic, and also because the more we can follow the same structure on multiple platforms when there's not a strong argument not to, the less cognitive load there is for us to maintain all the implementations (and for clients to learn the different APIs).

In that model the subclasses would be on the view, rather than the engine.

Reply:
Agreed, these certainly don't need to be public and I was wondering the best way to hide them from the API.

Reply:
I've updated this branch to have a FlEngine object that is public that contains a FlEngineImpl object that is private. This hides all the details of the engine, with public methods that we need able to be added to FlEngine. An API user can't easily tell if the FlEngine is X11/GL/Wayland etc, but I'm guessing this is probably not necessary and we can add methods like `fl_engine_is_x11` if necessary.